### DISCUSSION TITLE
Typing support

### DISCUSSION THREAD

**Discussion by pp-mo** [SciTools/iris#4383] [Ideas]
I reckon that [this line](https://github.com/SciTools/iris/blob/f696e5a9a50ac91572a1a4dcbe90add8607f3328/lib/iris/tests/__init__.py#L882) is our first use of type-hints in Iris code (:+1: 🚀 @jamesp ).
This is consistent with [dropping support for Python3.6](https://scitools-iris.readthedocs.io/en/latest/whatsnew/3.0.html?highlight=3.6#v3-0-4-22-july-2021)

What do we feel about this, is there a case for adding type hints to major sections of the API ?
Perhaps we should also [make a statement in the 3.2 whatsnew](https://github.com/SciTools/iris/issues/4382)

**Comment by trexfeathers** [comment]
I would imagine adding as I go, in the same way as I do with Numpy docstrings (which contradicts #3840 but we just haven't found time yet).

**Comment by jamesp** [comment]
@pp-mo see also #4272. Some great input from Jon Seddon

**Comment by jamesp** [comment]
Also, while type hints were not "formalised" in some sense until relatively recently, I think that annotations and the typing library preceed python 3.6. Adding type annotations should not cause issues with running on 3.6 if that were necessary, they are not interpreted at runtime.

Edit: Terrible grammar

**Comment by bjlittle** [comment]
Just had a lovely typing conversation with @rcomer regarding #4294, where being a very good `iris` citizen, she added type annotations to her PR.

However, this conversation raised quite a lot of open questions that really do need a group steer to make the process of adding typing in `iris` way, way less painful i.e., IMHO we really do need to start to curate an evolving typing best practice

1. Bootstrapping typing stub files seems like the scalable and maintainable way to go. `iris` and it's data model is complex, so it would be ideal to start simple with common `iris` typing axioms and build/scale outwards
2. Dealing with arrays is just confusing (due to a lack of typing maturity atm even with `numpy.typing`?). We need clarity on generic typing and when to discriminate between specific real (`numpy`), lazy (`dask`), `ndarray` and `masked` cases
3. When is it best to specialise with specific array `dtype` and `shape`?
4. What audience is typing targeting? Users? Developers? or both? i.e., this leads to the question of type hinting private classes, functions and methods
5. We need clarity on when to use `collections.abc.Sequence` vs `collections.abc.Iterable` vs `collections.abc.Iterator` vs `typing.Iterable`
6. We need to know the pros and cons of when to use a typing alias vs a `typing.TypeVar`
7. Typing the code is only half of the problem. How the typing is rendered in documentation is the other. It's very easy to create a typing documentation abomination :scream: 
8. Typing can render documentation and doc-strings hard / very difficult to read. Somehow `xarray` appear to have somewhat squared that circle, see https://xarray.pydata.org/en/v0.13.0/generated/xarray.merge.html (thanks to @jamesp for sharing this observation)

Just thought I'd openly capture and share our conversation questions here... as I can only assume it might resonate for anyone else who has dipped their toe into the typing world for ~~bloated~~ large legacy code bases.

Have I missed or misrepresented anything @rcomer ?

Ideally, it would be great to agree on a way forward using #4294 as a concrete use case and roll from there...

**Reply by rcomer** [reply]
Thanks @bjlittle, that's a great summary.  I think the only other thing on my list was the rather specific question of how to hint for something that could be a built-in or numpy type (but basically the same thing).  E.g. where the code clearly expects that the variable [may be an int or np.integer](https://github.com/SciTools/iris/blob/1093f711cd45339fafc34447952cfb19c703e838/lib/iris/coords.py#L1953) is it enough to just hint that it's an `int`?

**Reply by rcomer** [reply]
Typing changes in #4294 are currently contained in the [last 4 commits](https://github.com/SciTools/iris/pull/4294/files/63b372c0a612ebca961716ac20b8b4ce36988b96..0361de4fa8fc97514bccec13d6088fa82e8c318b).

**Comment by bjlittle** [comment]
@jamesp also shared this link, which is a great read, see https://sethmlarson.dev/blog/2021-10-18/tests-arent-enough-case-study-after-adding-types-to-urllib3

**Comment by rcomer** [comment]
For the rendered docs, it looks like `xarray` have simply turned off the type hints:
https://github.com/pydata/xarray/blob/e056cacdca55cc9d9118c830ca622ea965ebcdef/doc/conf.py#L112
https://www.sphinx-doc.org/en/master/usage/extensions/autodoc.html#confval-autodoc_typehints

The advantages of typing discussed at #4272 seem to focus on how editors can interpret them, rather than documentation.  I'm particularly not keen on how `Optional` is shown as the typing `Optional` is specifically [not the same thing as an optional argument](https://docs.python.org/3/library/typing.html#typing.Optional), which seems unintuitive to me.  Would much rather it said "or None"
![collapsed](https://user-images.githubusercontent.com/10599679/148365430-f9eddf86-d8b7-424d-a187-9e756f3a2002.png)

So, if we're voting, I think I'll vote for turning them off in the docs.

**Reply by bjlittle** [reply]
Yeah, point [8] from [comment above](https://github.com/SciTools/iris/discussions/4383#discussioncomment-1913397) is a BIG concern for me atm

**Reply by jamesp** [reply]
@rcomer I'd agree with turning off in the docs, the default rendering of the Python type objects is horrid.  xarray does have some affordance for types - see the links for iterable, DataArray etc in the Parameters section below the signature. But that seems to come from a clever parsing of the docstring, not the types so requires a certain amount of duplication of information.

![image](https://user-images.githubusercontent.com/22805/148366722-d7476e2e-532f-47f2-bb74-b0c2987fba1b.png)

However, I'd argue that expecting a type system to render to nice documentation automatically is going to be unlikely.  The xarray example above is a good case.  The type signature of `objects` is `objects: Iterable[Union["DataArray", "CoercibleMapping"]]`.  What is a CoercibleMapping?!  The doc string is still much more relevant to the user, and therefore more appropriate in the API docs.

**Reply by trexfeathers** [reply]
I can see the logic to this. Although it does of course mean that we need to remember to keep including the types in the docstring as well as the signature (not needing to do this was something @pp-mo was looking forward to!).

**Reply by pp-mo** [reply]
>  keep including the types in the docstring as well as the signature

It's a shame for DRY (and while trialling typechecks elsewhere, I've already had problems with the two getting out-of-step)
But I totally agree that docs readability has to come first.

**Reply by rcomer** [reply]
PR to turn them off here: #4510

**Reply by pp-mo** [reply]
I think I may have found a "better way" for this using `autodoc_typehints = "description"`
See [comment here](https://github.com/SciTools/iris/pull/4510#issuecomment-1789348236)

**Reply by pp-mo** [reply]
I think some of the other negative behaviours noted above may have eased, too.

E.G. 
> @jamesp not keen on how Optional is shown ... Would much rather it said "or None"

I believe that is now fixed : napoleon now renders  `Optional(int)` as `int | None`.

**Comment by rcomer** [comment]
> We need clarity on when to use collections.abc.Sequence vs collections.abc.Iterable vs collections.abc.Iterator vs typing.Iterable

Today I learned that if you specify `thing` is an `Iterable`, PyRight complains if you do `len(thing)`.  So in this case `Sequence` appears to be the way to go.

[Edit: assuming PyRight is actually Right...]

**Reply by rcomer** [reply]
Although a set is an `Iterable` and has a length, but is not a `Sequence`...

**Reply by jamesp** [reply]
Maybe `Sized` is what you're looking for?

https://docs.python.org/3/library/typing.html#typing.Sized

**Reply by rcomer** [reply]
Hmmm.  I want to check the length _and_ loop through it...

**Reply by rcomer** [reply]
Maybe I want a `Collection`
https://docs.python.org/3/library/typing.html#typing.Collection

**Reply by rcomer** [reply]
👀 handy table here

https://docs.python.org/3/library/collections.abc.html#collections-abstract-base-classes

**Comment by trexfeathers** [comment]
Closing following the creation of #5924 - we do want to do this one way or another
